perlop: Fixups

khwilliamson opened the pull request:
Fixes #18565

* This set of changes does not require a perldelta entry.

book commented on pod/perlop.pod:
```diff
@@ -1517,11 +1519,22 @@ takes higher precedence.
+X<operator, logical, xor> X<operator, logical, exclusive or> X<xor>
+
+Binary C<"xor"> returns the logical disjunction of the two surrounding
+expressions.  That means it returns 1 if either, but not both, are true.
```
Does it return `1` or `true` (the boolean)?

khwilliamson replied:
`true`.  Fixed

happy-barney commented on pod/perlop.pod:
```diff
@@ -1517,11 +1519,22 @@ takes higher precedence.
+and not match, for otherwise it would be a bug in our pattern matching
+code.
+
+ ($a =~ /$pat/ xor $a !~ /$pat/) or die;
```
with empty regex `//` to be same as defined-or operator `//` it will be nice to not promote operator-less regular expressions:
```
$a =~ $pat
$a =~ qr /$pat/
```

khwilliamson replied:
Thanks; now fixed

Grinnz commented on pod/perlop.pod:
```diff
@@ -3836,6 +3849,11 @@ Here is a short, but incomplete summary:
 
 Choose wisely.
 
+=head2 Logical or and Exclusive Or
```
I believe these were put in the same section because the current organization of perlop is to document operators in the order of precedence, with operators of the same precedence (like or and xor) sharing a section.

khwilliamson replied:
Is it then the case that there is a distinct precedence order in ``&& || ^^ // ..`` as these all have their own section in this order?  My guess is that ``||`` and ``^^`` have the same precedence.

Grinnz replied:
According to the precedence table at the top, `&&` has a distinct precedence and `||` `//` `^^` share a precedence. Perhaps this was an oversight when adding the latter two operators.

Grinnz commented on pod/perlop.pod:
```diff
@@ -1488,14 +1492,12 @@ expressions.  It's equivalent to C<&&> except for the very low
-Binary C<"or"> returns the logical disjunction of the two surrounding
-expressions.  It's equivalent to C<||> except for the very low precedence.
-This makes it useful for control flow:
+Logical C<"or"> returns the logical inclusive disjunction of the two
```
I do not think "binary" here refers to the operation, but to the syntax of the operator: it takes two arguments. Note that the versions of these operators marked as "bitwise" as well as the C-style logical forms are also introduced with this term. In fact it is distinguishing for example binary `-` with unary `-`.

khwilliamson replied:
You're right about the intent, based on looking at other portions of the document.  But it confused me, and thus could easily confuse others; so some clarification is necessary.

Grinnz reviewed (dismissed):
I think or and xor should be kept as the same section, and the "binary" term should not be changed, otherwise LGTM

Grinnz commented:
> Also that I had wrongly said `xor` is the equivalent of `^`. It instead is the equivalent of `^^` (which I hadn't known existed).

It had not existed, until the 5.40 release. https://perldoc.perl.org/perl5400delta#New-%5E%5E-logical-xor-operator

Grinnz commented on pod/perlop.pod:
```diff
@@ -1517,11 +1519,22 @@ takes higher precedence.
+expression pattern can't both match and not match, for otherwise it
+would be a bug in our pattern matching code.
+
+ ($a =~ qr/$pat/ xor $a !~ qr/$pat/) or die;
```
Avoid promoting use of `$a` in docs
```suggestion
 ($x =~ qr/$pat/ xor $x !~ qr/$pat/) or die;
```

khwilliamson replied:
agreed

khwilliamson committed:
perlop: Update advice on relational ops and Unicode

Enhancements have been made in the last few releases to using these
operators on UTF-8 locales; but this advice was not updated, so until
this commit, said don't do that.

khwilliamson commented:
It turns out that there a bunch of infelicitous things in this pod; so I'm working on a more extensive fix

khwilliamson committed:
perlop: Move isa() section to proper place

This document is supposed to be in decreasing operator precedence order.
isa() is higher than the relational operators, so move it to there.

khwilliamson committed:
perlop: Change 'if(' to 'if ('

It is more customary to separate an 'if' from its paren.

khwilliamson committed:
perlop: Fix grammatical nits

khwilliamson committed:
perlop: Move trivial case to prominence

Otherwise this dangling sentence gets lost after the main part of the
section.

khwilliamson committed:
Move discussion of relational ops into proper section

These paragraphs were separated from the main area where all but one of
their subjects, 'cmp',  were discussed.  Move them, and provide a link
to the outlier, and back to the main discussion.

khwilliamson committed:
perlop: Add some C< >.

khwilliamson committed:
perlop: Fix up smartmatch section

This has the same precedence as other operators, so should not be in a
separate section with the same heading level.

Solve this by making it at a sublevel.  This fits nicely with another
sublevel section on smartmatch that follows immediately.  There are now
to subsections.

khwilliamson committed:
perlop: Fix up a couple X<>

This gets things more rationally ordered

khwilliamson committed:
perlop: Use formal terminology to distinguish or, xor

khwilliamson committed:
perlop: Add detail about xor

Fixes #18565

khwilliamson committed:
perlop: Move discussion into proper sections

This detail for two sections was combined in a whole other section.
This splits the detail into text appropriate for each section to which
it applies, and moves it there.

khwilliamson committed:
perlop: Combine equal precedence ops into same section

The design of this pod is each section is for all operators of the same
precedence.  This was being violated

khwilliamson committed:
perlop: Specifically give an operations' results

khwilliamson committed:
perlop: A few clarifications

khwilliamson committed:
perlop: Specifically mention this document's structure

khwilliamson commented:
I found a bunch of problems/improvements to be made to this pod

I wonder if the smartmatch text needs updating.

Grinnz commented on pod/perlop.pod:
```diff
@@ -1014,24 +1022,52 @@ if the left operand is false, the right operand is not even evaluated.
+C<and> operator (see L<below|/Logical And>).
+The short-circuit behavior is identical.  The precedence of C<"and"> is
+much lower, however, so that you can safely use it after a list operator
+without the need for parentheses:
```
Ends in a colon but there is nothing following

Grinnz commented on pod/perlop.pod:
```diff
@@ -1489,13 +1502,14 @@ precedence.  This means that it short-circuits: the right
+X<operator, logical, or> X<or>
+X<operator, logical, xor> X<operator, logical, exclusive or> X<xor>
+
+There is no low precedence operator for defined-OR.
```
This feels slightly better at the end of this section, as an "addendum," in my opinion at least

khwilliamson replied:
How would that read?  I think the current situation is that it gets lost

Grinnz replied:
My feeling is that people would more likely go to this section of the documentation to find out about the or and xor operators, more than to search for a low precedence defined-or, but I have no objections to putting that sentence first either.

khwilliamson replied:
@Grinnz, I agree that that is why they would come here; but I moved the sentence because I thought that point would be lost where it was, and it's not really a problem to have it moved.

Now you say you don't object to moving the sentence.  What changes need to be made, then?

Grinnz replied:
Totally up to you, just giving my opinion on it

khwilliamson replied:
I'm open to suggestions for a better way

Grinnz commented:
> I wonder if the smartmatch text needs updating.

It should be updated since it is no longer considered experimental, and to describe the current state of the operator, which is that it is discouraged and unlikely to receive any further development.

khwilliamson commented:
No better ideas have been offered; so merging
